### PR TITLE
 Fix multiple bugs with progress deadline handling

### PR DESCRIPTION
**Pull request by dadgar**
Fix an issue in which the deployment watcher would fail the deployment
based on the earliest progress deadline of the deployment regardless of
if the task group has finished.

The PR also makes handling deployment status updates from the client more robust.

### CODE REVIEW

File: nomad/deploymentwatcher/deployment_watcher.go [hunk 1]
```diff
@@ -419,7 +417,12 @@ FAIL:
 					default:
 					}
 				}
-				deadlineTimer.Reset(next.Sub(time.Now()))
+
+				// If the next deadline is zero, we should not reset the timer
```
Comment (preetapan) [inline nomad/deploymentwatcher/deployment_watcher.go]: Could this clarify what problem this !zero check is actually fixing. i'm concerned that this means under some conditions the deadlinetimer would never get reset

File: nomad/deploymentwatcher/deployment_watcher.go [hunk 2]
```diff
@@ -611,16 +614,72 @@ func (w *deploymentWatcher) shouldFail() (fail, rollback bool, err error) {
...
+
+	allocs, err := snap.AllocsByDeployment(nil, d.ID)
+	if err != nil {
+		return nil
+	}
+
+	// Go through the allocs and count up how many healthy allocs we have
+	healthy := make(map[string]int, len(d.TaskGroups))
+	for _, a := range allocs {
+		if a.TerminalStatus() || !a.DeploymentStatus.IsHealthy() {
```
Comment (preetapan) [inline nomad/deploymentwatcher/deployment_watcher.go]: what about an alloc from a batch job that completed early? That should still count as healthy, but this code ignores terminal allocs

File: nomad/deploymentwatcher/deployment_watcher.go [hunk 3]
```diff
@@ -820,5 +871,10 @@ func (w *deploymentWatcher) jobEvalStatus() (latestIndex uint64, blocked bool, e
 		}
 	}
 
-	return max, false, nil
+	if max == uint64(0) {
```
Comment (preetapan) [inline nomad/deploymentwatcher/deployment_watcher.go]: This index returns the max eval index across all jobs, but we care about a single job. We could miss an update if evals were gced before this ran. This should return zero instead.

**Comment by dadgar** [comment]
@preetapan The server was actually panicking if the client sent an update without the deployment status set, so the new code, is more correct and safer.

**Pull request merged**